stdlib/public/Concurrency/TaskGroup.swift

```diff
@@ -228,21 +261,67 @@ extension Task {
     /// - SeeAlso: `Task.checkCancelled`
     /// - SeeAlso: `Task.isCancelled`
     public mutating func cancelAll() {
-      _taskCancel(self.task) // TODO: do we also have to go over all child tasks and cancel there?
+      _taskGroupCancelAll(task: _task, group: _group)
+    }
+
+    // Returns `true` if the group was cancelled, e.g. by `cancelAll`.
+    public var isCancelled: Bool {
+      return _taskIsCancelled(_task) ||
+        _taskGroupIsCancelled(task: _task, group: _group)
+    }
+  }
+}
+
+/// ==== TaskGroup: AsyncSequence.AsyncIterator --------------------------------
+
+extension Task.Group {
+
+  public func makeAsyncIterator() -> GroupIterator {
+    return GroupIterator(group: self)
+  }
+
+  /// Allows iterating over results of tasks added to the group.
+  ///
+  /// The order of elements returned by this iterator is the same as manually
+  /// invoking the `group.next()` function in a loop, meaning that results
+  /// are returned in *completion order*.
+  ///
+  /// - SeeAlso: `TaskGroup.next()`
+  public struct GroupIterator: AsyncIteratorProtocol {
+    public typealias Element = TaskResult
+
+    @usableFromInline
+    var group: Task.Group<TaskResult>
+
+    // no public constructors
+    init(group: Task.Group<TaskResult>) {
+      self.group = group
+    }
+
+    /// - SeeAlso: `TaskGroup.next()` for a detailed discussion its semantics.
+    public mutating func next() async throws -> Element? {
+      try await group.next()
+    }
+
+    public mutating func cancel() {
+      group.cancelAll()
```

Comment:
TODO: this may be slightly wrong; add a test to throw out of a loop, but catch it -- the group should remain NOT cancelled as we ONLY cancel the group if we throw out of the group, not just out of the next()